vignettes/test-fixtures.Rmd

```diff
@@ -31,21 +31,62 @@ Ideally, a test should leave the world exactly as it found it. But you often nee
+This article introduces a powerful technique that allows you to solve both problems: **test fixtures**. We'll begin by discussing some pre-existing tools, then learn about the tools that make fixtures possible, talk about exactly what a test fixture is, and show a few examples.
 
 Much of this vignette is derived from <https://www.tidyverse.org/blog/2020/04/self-cleaning-test-fixtures/>; if this is your first encounter with `on.exit()` or `withr::defer()`, I'd recommend starting with that blog as it gives a gentler introduction. This vignette moves a little faster since it's designed as more of a reference doc.
 
 ```{r}
 library(testthat)
 ```
 
+## `local_` helpers
+
+We'll begin by giving you the bare minimum of knowledge to change global state _just_ within your test. The withr package provides a number of functions that temporarily change the state of the world, carefully undoing the changes when the current function finishes:
+
+| Do / undo this              | withr function    |
+|-----------------------------|-------------------|
+| Create a file               | `local_tempfile()`|
+| Create a directory          | `local_tempdir()` |
+| Set an R option             | `local_options()` |
+| Set an environment variable | `local_envvar()`  |
+| Change working directory    | `local_dir()`     |
+
+(You can see a full list at <https://withr.r-lib.org/> but these five are by far the most commonly used.)
+
+These allow you to control options that would otherwise be painful. For example, imagine you were testing the base R code that rounds numbers to a fixed number of places when printing. You could write code like this:
+
+```{r}
+test_that("print() respects digits option", {
+  x <- 1.23456789
+
+  withr::local_options(digits = 1)
+  expect_equal(capture.output(x), "[1] 1")
+
+  withr::local_options(digits = 5)
+  expect_equal(capture.output(x), "[1] 1.2346")
+})
+```
+
+If you write a lot of code like this in your tests, you might decide you want a helper function or **test fixture** that reduces the duplication. Fortunately withr's local functions us to solve this problem by providing an `.local_envir` or `envir` argument that controls when cleanup occurs. The exact details of how this works are rather complicated, but fortunately there's a common pattern you can use without understanding all the details. Your helper function should always have an `env` argument that defaults to `parent.frame()`, which you pass to the `.local_envir` argument of `local_()`:
+
+```{r}
+local_digits <- function(sig_digits, env = parent.frame()) {
+  withr::local_options(digits = sig_digits, .local_envir = env)
+
+  # mark that this function is called for its side-effects not its return value
+  invisible() 
+}
+```
+
+This is particularly useful when you need to chain together multiple `local_` functions to achieve the desired state.
```

Comment:
This feels like some text got lifted from elsewhere or moved and it doesn't quite make sense anymore? It doesn't feel like the first example, using `local_options()` segues naturally into the 2nd paragraph ("If you write a lot of code like this ..."). What "duplication" are we talking about? It feels like this must have unfolded differently in the original prose? Sorry this is a vague impression.

Reply:
Removed, thanks!